src/hexkit/log.py

```diff
@@ -0,0 +1,160 @@
+            + " following can also be specified: timestamp, service, instance, level,"
+            + " correlation_id, and details"
+        ),
+    )
+
+
+class JsonFormatter(Formatter):
+    """A formatter class that outputs logs in JSON format."""
+
+    def format(self, record: LogRecord) -> str:
+        """Format the specified record as a JSON string.
+
+        This will format the log record as JSON with the following values (in order):
+            - timestamp: The ISO 8601-formatted timestamp of the log message.
+            - service: The name of the service where the log was generated.
+            - instance: The instance ID of the service where the log was generated.
+            - level: The log's severity.
+            - name: The name of the logger.
+            - correlation_id: The correlation ID, if set, from the current context.
+            - message: The message that was logged, formatted with any arguments.
+            - details: Any additional values included at time of logging.
+        """
+        # Create a log record dictionary
+        log_record = record.__dict__
+        output: OrderedDict[str, str] = OrderedDict()
+
+        output["timestamp"] = log_record["timestamp"]
+        output["service"] = log_record["service"]
+        output["instance"] = log_record["instance"]
+        output["level"] = log_record["level"]
+        output["name"] = log_record["name"]
+        output["correlation_id"] = log_record["correlation_id"]
+        output["message"] = record.getMessage()  # construct msg str with any args
+        output["details"] = log_record["details"]
+
+        # Convert to JSON string
+        return json.dumps(output)
+
+
+class RecordCompiler(StreamHandler):
+    """A class to make all non-standard information available to formatters."""
+
+    def __init__(self, *, config: LoggingConfig):
+        """Initialize with logging config."""
+        super().__init__()
+
+        self._service_name = config.service_name
+        self._service_instance_id = config.service_instance_id
+
+    def handle(self, record: LogRecord) -> bool:
+        """Set custom record attributes"""
+        log_record = record.__dict__
+        timestamp = datetime.fromtimestamp(log_record["created"])
+        timestamp = timestamp.astimezone(timezone.utc)
+        iso_timestamp = timestamp.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
+        extras = {
+            key: value
+            for key, value in log_record.items()
+            if key not in RESERVED_RECORD_KEYS
+        }
+        record.timestamp = iso_timestamp
+        record.service = self._service_name
+        record.instance = self._service_instance_id
+        record.level = log_record["levelname"]
+        record.correlation_id = correlation_id_var.get(None)
+        record.details = extras
+        return super().handle(record)
+
+
+def configure_logging(*, logger: Logger, config: LoggingConfig):
+    """Set up logging"""
```

Comment:
Actually, it might be nicer to use the root logger by default, because this is most likely what we want to do to enable logging done by other libraries:
```suggestion
def configure_logging(*, config: LoggingConfig, logger: Optional[Logger] = None):
    """Configure logging for the specified logger. If no logger is provided, the root logger is use."""
    if not logger:
        logger = logging.getLogger()
    
```

Reply:
I agree that it would be easier to get 3rd party loggers to work that way. Please see [a4e2c2d](https://github.com/ghga-de/hexkit/pull/77/commits/a4e2c2da6dca7ccd3144a9aa436aa266cadc1d00) for relevant changes